### PR TITLE
feat: [FC-0047] Relative Dates

### PR DESCRIPTION
**Pull request by IvanStepanok**
### New Feature: Relative Date Display
The useRelativeDates functionality has been added to improve the readability of dates within the application. When this feature is enabled (default setting), dates within the range of -7 to +7 days from the current date will be displayed in a relative format (e.g., "just now," "2 days ago"). Dates outside this range will continue to be displayed in the traditional format. This setting can be toggled in the calendar synchronization settings.

### Quick Demo

https://github.com/user-attachments/assets/9ded7849-ed3d-4b69-8f74-7cd7723cd45e

### Environment for testing:
```
API_HOST_URL: 'https://axim-mobile.raccoongang.net'
ENVIRONMENT_DISPLAY_NAME: 'lms-axim-stage'
FEEDBACK_EMAIL_ADDRESS: 'support@example.com'
OAUTH_CLIENT_ID: 'zP3vPz00c8fTRpYjNbVSlA1fxt9LnCxTM4JK1KQ0'
DISCOVERY:
    TYPE: "native"
    
FIREBASE:
    ENABLED: false

UI_COMPONENTS:
    COURSE_UNIT_PROGRESS_ENABLED: false
    COURSE_NESTED_LIST_ENABLED: false

```

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @IvanStepanok!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.




#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/openedx-mobile-maintainers`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

**Comment by sdaitzman** [comment]
Hi @IvanStepanok, once the date handling has been updated to match the comment above/what @PavloNetrebchuk [described](https://github.com/openedx/openedx-app-android/pull/367#issuecomment-2249995923) on the Android version, I think this is all clear from the design side for now.

In the future, I think it's worth coming back to this and reconsidering how we approach dates/examining some edge cases, so I've created a new roadmap issue (#513) to track that future work.

File: Dashboard/Dashboard/Presentation/Elements/PrimaryCardView.swift
Comment (salman2013) [inline Dashboard/Dashboard/Presentation/Elements/PrimaryCardView.swift:275]: This can be in one line

File: Discovery/Discovery/Presentation/NativeDiscovery/DiscoveryViewModel.swift
Comment (salman2013) [inline Discovery/Discovery/Presentation/NativeDiscovery/DiscoveryViewModel.swift:40]: Any specific reason to make it public? I thought we can create getter of this property.

File: Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift
Comment (salman2013) [inline Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift:62]: We should use weak self inside closures like [weak self]

File: Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift
Comment (salman2013) [inline Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift:64]: We should use weak self inside closures like [weak self]

File: Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift
Comment (salman2013) [inline Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift:113]: We should use weak self inside closures like [weak self]

File: Discussion/DiscussionTests/Presentation/Comment/Base/BaseResponsesViewModelTests.swift
Comment (salman2013) [inline Discussion/DiscussionTests/Presentation/Comment/Base/BaseResponsesViewModelTests.swift:411]:
```
        let interactor = DiscussionInteractorProtocolMock()
        let router = DiscussionRouterMock()
        let config = ConfigMock()
        let viewModel = BaseResponsesViewModel(
            interactor: interactor,
            router: router,
            config: config,
            storage: CoreStorageMock()
        )
```

These properties are using in multiple methods so we should define them as instance properties of the test class and initialize them in the `setUp` method.

e.g

`override func setUp() {
        super.setUp()
        
        interactor = DiscussionInteractorProtocolMock()
        router = DiscussionRouterMock()
        config = ConfigMock()
        viewModel = BaseResponsesViewModel(
            interactor: interactor,
            router: router,
            config: config,
            storage: CoreStorageMock()
        )
    }`

File: Discussion/DiscussionTests/Presentation/Posts/PostViewModelTests.swift
Comment (salman2013) [inline Discussion/DiscussionTests/Presentation/Posts/PostViewModelTests.swift:188]: As per my previous comment, the object is used in multiple methods so we can initialize it in the `setup` method and use it in the whole class in numerous methods.

File: OpenEdX/DI/ScreenAssembly.swift
Comment (salman2013) [inline OpenEdX/DI/ScreenAssembly.swift:191]: I don't know the reason for forced unwrapping here but it can be dangerous and can produce crashes.

Comment (IvanStepanok) [reply OpenEdX/DI/ScreenAssembly.swift:191]: Thank you for your observation. While your point about the dangers of forced unwrapping is theoretically valid, in practice, it is crucial for us that the application crashes if Swinject Dependency Injection fails. Essentially, if the DI container fails to resolve a dependency, the application would be non-functional, and it's better to know this immediately through a crash rather than running with missing dependencies. This ensures that we address the issue promptly and maintain the integrity of the application's behavior.

File: Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift
Comment (IvanStepanok) [reply Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift:62]: Thank you for your suggestion. In this particular case, using weak self is not possible because ResponsesView is not a class or a class-bound protocol. The weak keyword can only be applied to class instances or class-bound protocols. Therefore, it is not applicable to structs or other data types like ResponsesView. I hope this clarifies the situation.

File: Discovery/Discovery/Presentation/NativeDiscovery/DiscoveryViewModel.swift
Comment (IvanStepanok) [reply Discovery/Discovery/Presentation/NativeDiscovery/DiscoveryViewModel.swift:40]: Thank you for your question. The reason for making `connectivity` public is because it is directly used inside the view. While it is possible to create a public getter and keep the property private, making the property itself public simplifies access in this case, especially since there are no setters and the property is effectively read-only. This ensures that the view has direct and straightforward access to the property without additional boilerplate code.

When we started the project, we decided to follow this approach for consistency. Although your point is valid, it is not critical in this context, and to maintain a uniform coding style, it is better to stick to the initially chosen method.

File: Dashboard/Dashboard/Presentation/Elements/PrimaryCardView.swift
Comment (IvanStepanok) [reply Dashboard/Dashboard/Presentation/Elements/PrimaryCardView.swift]: Thanks!

File: Core/Core/Extensions/DateExtension.swift
Comment (IvanStepanok) [reply Core/Core/Extensions/DateExtension.swift]: Make sense

**Comment by IvanStepanok** [comment]
@salman2013 Thanks for your review; all feedback has been addressed.

File: Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift
Comment (salman2013) [reply Discussion/Discussion/Presentation/Comments/Responses/ResponsesView.swift:62]: Yes, you are right i read it as a class.

File: OpenEdX/DI/ScreenAssembly.swift
Comment (salman2013) [reply OpenEdX/DI/ScreenAssembly.swift:191]: Make sense for now, i would suggest using `guard` here in future with some error alert instead of crashing the app.

**Review by salman2013 — APPROVED** [review]
It looks good to me.

**Review by volodymyr-chekyrta — APPROVED** [review]

**Review by volodymyr-chekyrta — APPROVED** [review]

**Comment by openedx-webhooks** [comment]
@IvanStepanok 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-app-ios&entry.867055334=https://github.com/openedx/openedx-app-ios/pull/505&entry.1484655318=https://github.com/IvanStepanok&entry.752974735=2024-07-19+08:31&entry.1917517419=2024-08-23+08:11&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
